File: ts/tests/nested.test.ts
// SPDX-FileCopyrightText: 2026 Blackcat Informatics Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: MIT OR Apache-2.0

import { readFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { test } from "node:test";
import assert from "node:assert/strict";

import type { Diagnostic } from "../src/model.js";
import { readNested } from "../src/nested.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const repoRoot = resolve(__dirname, "../../../");

interface NestedSecurityVector {
    id: string;
    max_depth: number;
    expected_diagnostics: string[];
}

function hasDiagnostic(diagnostics: Diagnostic[], code: string): boolean {
    return diagnostics.some((diag) => diag.code === code);
}

function readHexFixture(name: string): Uint8Array {
    const hex = readFileSync(
        join(repoRoot, "vectors", "security", name),
        "utf-8",
    ).replace(/\s+/g, "");
    return Uint8Array.from(Buffer.from(hex, "hex"));
}

function loadVector(): NestedSecurityVector {
    return JSON.parse(
        readFileSync(
            join(
                repoRoot,
                "vectors",
                "security",
                "nested-recursion-limit.json",
            ),
            "utf-8",
        ),
    ) as NestedSecurityVector;
}

test("readNested exposes subgraphs by blob digest", () => {
    const result = readNested(
        readHexFixture("nested-recursion-limit.gts.hex"),
        {
            maxDepth: 3,
            maxDecodedBytes: 16 * 1024 * 1024,
        },
    );
    const child = result.graph.blobs[0];

    const subgraph = result.subgraph(child.digest);
    assert.ok(subgraph);
    assert.equal(subgraph.blobs.length, 1);
    assert.equal(hasDiagnostic(result.diagnostics, "RecursionLimit"), false);
});

test("nested-recursion-limit security vector", () => {
    const vector = loadVector();
    const result = readNested(
        readHexFixture("nested-recursion-limit.gts.hex"),
        {
            maxDepth: vector.max_depth,
            maxDecodedBytes: 16 * 1024 * 1024,
        },
    );

    const child = result.graph.blobs[0];
    const childGraph = result.subgraph(child.digest);
    assert.ok(childGraph);
    const grandchild = childGraph.blobs[0];
    assert.equal(result.subgraph(grandchild.digest), undefined);
    for (const code of vector.expected_diagnostics) {
        assert.equal(hasDiagnostic(result.diagnostics, code), true);
    }
});

test("readNested stops at decoded-size budget", () => {
    const result = readNested(
        readHexFixture("nested-recursion-limit.gts.hex"),
        {
            maxDepth: 3,
            maxDecodedBytes: 0,
        },
    );

    assert.equal(result.subgraphs.size, 0);
    assert.equal(hasDiagnostic(result.diagnostics, "RecursionLimit"), true);
});

test("readNested skips duplicate nested blob digests", () => {
    const fixture = readHexFixture("nested-duplicate-digest.gts.hex");
    const baseline = readNested(fixture, {
        maxDepth: 3,
        maxDecodedBytes: 16 * 1024 * 1024,
    });
    const childA = baseline.graph.blobs[0];
    const childB = baseline.graph.blobs[1];
    const childAGraph = baseline.subgraph(childA.digest);
    assert.ok(childAGraph);
    const sharedGrandchild = childAGraph.blobs[0];
    const exactBudget =
        childA.data.length + childB.data.length + sharedGrandchild.data.length;

    const result = readNested(fixture, {
        maxDepth: 3,
        maxDecodedBytes: exactBudget,
    });

    assert.equal(result.subgraphs.size, 3);
    assert.equal(hasDiagnostic(result.diagnostics, "RecursionLimit"), false);
});

File: go/nested/nested_test.go
// SPDX-FileCopyrightText: 2026 Blackcat Informatics Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: MIT OR Apache-2.0

package nested

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"go.blackcatinformatics.ca/gts/model"
	"go.blackcatinformatics.ca/gts/wire"
	"go.blackcatinformatics.ca/gts/writer"
)

const ex = "https://example.org/"

type nestedSecurityVector struct {
	ID                  string   `json:"id"`
	MaxDepth            int      `json:"max_depth"`
	ExpectedDiagnostics []string `json:"expected_diagnostics"`
}

func iri(value string) model.Term {
	return model.Term{Kind: model.Iri, Value: value}
}

func lit(value string) model.Term {
	return model.Term{Kind: model.Literal, Value: value}
}

func tinyGraph(label string) []byte {
	w := writer.New("dist")
	w.AddTerms([]model.Term{
		iri(ex + label),
		iri(ex + "label"),
		lit(label),
	})
	w.AddQuads([]model.Quad{{S: 0, P: 1, O: 2}})
	return w.ToBytes()
}

func bundle(child []byte) []byte {
	w := writer.New("bundle")
	w.AddBlob(child, GTSMediaType, "")
	return w.ToBytes()
}

func labeledBundle(child []byte, label string) []byte {
	w := writer.New("bundle")
	w.AddTerms([]model.Term{
		iri(ex + label),
		iri(ex + "label"),
		lit(label),
	})
	w.AddQuads([]model.Quad{{S: 0, P: 1, O: 2}})
	w.AddBlob(child, GTSMediaType, "")
	return w.ToBytes()
}

func hasDiagnostic(diags []model.Diagnostic, code string) bool {
	for _, diag := range diags {
		if diag.Code == code {
			return true
		}
	}
	return false
}

func loadNestedSecurityVector(t *testing.T) nestedSecurityVector {
	t.Helper()
	path := filepath.Join("..", "..", "vectors", "security", "nested-recursion-limit.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	var vector nestedSecurityVector
	if err := json.Unmarshal(data, &vector); err != nil {
		t.Fatal(err)
	}
	return vector
}

func TestReadNestedExposesSubgraphByBlobDigest(t *testing.T) {
	child := tinyGraph("child")
	outer := bundle(child)

	result := ReadNested(outer, 3, 16*1024*1024)

	subgraph, ok := result.Subgraph(wire.DigestStr(child))
	if !ok {
		t.Fatalf("missing nested subgraph")
	}
	if len(subgraph.Quads) != 1 {
		t.Fatalf("expected one child quad, got %d", len(subgraph.Quads))
	}
	if hasDiagnostic(result.Diagnostics, "RecursionLimit") {
		t.Fatalf("unexpected RecursionLimit diagnostic")
	}
}

func TestNestedRecursionSecurityVector(t *testing.T) {
	vector := loadNestedSecurityVector(t)
	grandchild := tinyGraph("grandchild")
	child := bundle(grandchild)
	outer := bundle(child)

	result := ReadNested(outer, vector.MaxDepth, 16*1024*1024)

	if _, ok := result.Subgraph(wire.DigestStr(child)); !ok {
		t.Fatalf("first nested child should be exposed")
	}
	if _, ok := result.Subgraph(wire.DigestStr(grandchild)); ok {
		t.Fatalf("grandchild should be blocked by max depth %d", vector.MaxDepth)
	}
	for _, code := range vector.ExpectedDiagnostics {
		if !hasDiagnostic(result.Diagnostics, code) {
			t.Fatalf("missing expected diagnostic %q", code)
		}
	}
}

func TestReadNestedStopsAtDecodedSizeBudget(t *testing.T) {
	child := tinyGraph("oversized")
	outer := bundle(child)

	result := ReadNested(outer, 3, len(child)-1)

	if _, ok := result.Subgraph(wire.DigestStr(child)); ok {
		t.Fatalf("oversized child should not be exposed")
	}
	if !hasDiagnostic(result.Diagnostics, "RecursionLimit") {
		t.Fatalf("missing RecursionLimit diagnostic")
	}
}

func TestReadNestedSkipsDuplicateDigest(t *testing.T) {
	grandchild := tinyGraph("shared-grandchild")
	childA := labeledBundle(grandchild, "child-a")
	childB := labeledBundle(grandchild, "child-b")
	w := writer.New("bundle")
	w.AddBlob(childA, GTSMediaType, "")
	w.AddBlob(childB, GTSMediaType, "")

	result := ReadNested(w.ToBytes(), 3, len(childA)+len(childB)+len(grandchild))

	if len(result.Subgraphs) != 3 {
		t.Fatalf("expected three subgraphs with one shared grandchild, got %d", len(result.Subgraphs))
	}
	if hasDiagnostic(result.Diagnostics, "RecursionLimit") {
		t.Fatalf("duplicate digest should not consume the decoded-size budget twice")
	}
}
